<a href="https://colab.research.google.com/github/MarianaCM004/Analitica-de-Negocios/blob/main/RETO_1_CAR_CRASH.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**PLANTEAMIENTO:**

De acuerdo con la base de datos Car Crash (Apendice A, Conjunto 1: Datos de Accidentes Automovilísticos del Libro Guia Analítica de Negocios Comunicación con Datos), seleccionar los datos para el ***condado de San Francisco***. Predecir si un accidente automovilístico **se provocó en una autopista o no**(Highway).

Para la predicción de los accidentes, es importante tener en cuenta las siguientes variables: Dia de la Semana (WeekDay), Nivel de Impacto (ViolCat), Cantidad de Luz Diurna, Mes, Crash Type (Convertir a Valor Numérico según el Orden Alfabético) y Cielo Despejado. El orden de las variables en las columnas se debe mantener como esta en el archivo de Excel.

De acuerdo con la base de datos anterior, llevar a cabo en el análisis la descripción de cada una de las variables (tipo de datos, categorías), eliminando de plano los registros que poseen datos faltantes.   


**Paso 1:**
Importamos las librerias necesarias para realizar el ejercicio

In [ ]:
import pandas as pd #ESTA LIBRERIA TIENE LAS FUNCIONES NUMERICAS
import numpy as np #Esta libreria se encarga de conectarme el excel con el colab

**Paso 2:**
Subir y leer base de datos

In [ ]:
data = '/content/1. BD2_CarCrash_Int.xlsx'
XDB = pd.read_excel(data, sheet_name=0) #XDB=XDataBase

ID      County            City  Weekday  Severity  ViolCat  \
0            1   SAN DIEGO       SAN DIEGO        7         1        8   
1            2    HUMBOLDT  UNINCORPORATED        4         1        8   
2            3     VENTURA          OXNARD        2         1       12   
3            4  STANISLAUS  UNINCORPORATED        4         1        1   
4            5   MENDOCINO  UNINCORPORATED        5         1        1   
...        ...         ...             ...      ...       ...      ...   
112655  112656      SHASTA         REDDING        1         0       11   
112656  112657      SHASTA         REDDING        6         1       11   
112657  112658      SHASTA         REDDING        5         0       11   
112658  112659      SHASTA         REDDING        3         0       11   
112659  112660      SHASTA         REDDING        5         1       11   

        ClearWeather  Month CrashType  Highway  Daylight  
0                  0      1         A        0         1  
1                  1      1         A        1         0  
2                  1      2         A        0         1  
3                  1      1         A        0         0  
4                  1      1         A        1         0  
...              ...    ...       ...      ...       ...  
112655             1      3         G        0         1  
112656             1      2         G        0         0  
112657             1      5         G        0         1  
112658             1      7         G        0         1  
112659             1      8         G        1         0  

[112660 rows x 11 columns]

**Paso 3:**
Eliminar valores vacíos

In [ ]:
XDBClean = XDB.dropna() #Con esta formula podemos deshacernos de las filas/columnas que tengan datos faltantes

**Paso 4:**
Convertimos los valores de texto de la columna crashtype en valores numericos

In [ ]:
XDBClean["CrashType"] = pd.Categorical(XDBClean["CrashType"]).codes  #Con esta formula podemos convertir los valores de texto en numero del antes iban de A-G y ahora van de 0-6

***Paso 5:***
Seleccionamos los datos que nos sirven de entrada y de salida para el modelo. Filtramos datos que correspondan SOLO a San Francisco

In [ ]:
XDBClean= XDBClean[['County','Weekday','ViolCat','ClearWeather','Month','Daylight','CrashType','Highway']] #Cargamos las variables de interes
XDBB = XDBClean[XDBClean["County"] == "SAN FRANCISCO"] #Seleccionamos la columna county, y los datos que correspondan a San francisco.

**Paso 6:**
Crear dataframes separados para las entradas y salidas

In [ ]:
XD=XDBB[['Weekday','ViolCat','ClearWeather','Month','CrashType','Daylight']]
YD=XDBB['Highway']

**Paso 7:**
Aplicar modelo de Naive Bayes, importanto la función gausiana


In [ ]:
from sklearn.naive_bayes import GaussianNB #la funcion exponencial de VP=Exp (-1/2((x-4)/devisacion estandar)^2) que necesitamos es una funcion gausiana, por eso lo importamos.
mnb = GaussianNB () #mnb=model naive bayes
mnb.fit(XD, YD)

GaussianNB()

In [ ]:
ct=mnb.class_count_
print ("El total de los datos es:\n", len(XD)) #Contamos el total de los datos
print ("Como estan separados los datos\n", ct) #Separamos los datos en 0 y 1. 0 siendo NO y 1 siendo SÍ

conteo = YD.value_counts() #Calculamos el porcentaje al que corresponde 0 y 1 respecto al total de datos
porcentaje_categorias = (conteo/len(XD))*100
print(porcentaje_categorias)

El total de los datos es:
 2081
Como estan separados los datos
 [1580.  501.]
Highway
0    75.925036
1    24.074964
Name: count, dtype: float64


**PREGUNTAS A RESPONDER CON BASE EN EL MODELO PLANTEADO:**

¿Cuál es el número de datos total de la base de datos?

¿Cuál es el porcentaje de los datos que hacen parte de cada una de las categorías de clasificación?

Determinar los intervalos en los cual se encuentra la variación del día de la semana para cada una de las categorías de clasificación

Determinar los intervalos en los cual se encuentra la variación de la ViolCat para cada una de las categorías de clasificación

Determinar el intervalo en los cuales se encuentra la variación de la Categoría de accidentes para cada una de las categorías de clasificación

¿Cuál es la exactitud alcanzada por el modelo?
¿Cuál es la Especificidad alcanzada por el modelo?

**Indicar el intervalo en donde se ubica la cantidad de verdaderos negativo**

**Indicar el intervalo en donde se ubica la cantidad de verdaderos positivos**

¿Cuál es la tasa de error alcanzada por el modelo?

¿Cuál es la Sensibilidad alcanzada por el modelo?

¿Cuál es la Precisión Alcanzada por el Modelo?

¿Indicar la categoría a la que pertenece una persona con las siguientes características [5,1,1,3,4,0]?


**Paso 8:** sacamos la media, desviacion estandar, varianza, lim sup e infe de cada categoria de la base de datos.

In [ ]:
#Se obtiene la info crucial de cada categoria con highway como variable dependiente

u=mnb.theta_
print("La media de cada categoria es:\n", u)

std=mnb.var_ #esto e svarianza, que es la sdt elevada al cuadrado
std=np.sqrt(std) #esto es la desviacion estandar
print("La desviación estandar de cada categoria es:\n", std)

#Se identifican los valores superiores en ineferiores de los ingresos
print("El limite superior de los valores de cada categoria es:\n" , u+std)
print("El limite inferior de los valores de cada categoria es:\n" , u-std)

La media de cada categoria es:
 [[3.88417722 7.45696203 0.86265823 4.35189873 3.21772152 0.66265823]
 [3.9261477  4.1497006  0.79840319 6.69461078 2.27744511 0.62075848]]
La desviación estandar de cada categoria es:
 [[1.95062542 3.52136997 0.34420782 2.3433645  1.87371962 0.47280262]
 [2.01603723 2.16198902 0.40119266 3.44163641 1.05765448 0.48519831]]
El limite superior de los valores de cada categoria es:
 [[ 5.83480263 10.978332    1.20686605  6.69526323  5.09144114  1.13546084]
 [ 5.94218494  6.31168962  1.19959585 10.13624719  3.33509959  1.10595679]]
El limite inferior de los valores de cada categoria es:
 [[1.9335518  3.93559205 0.51845041 2.00853424 1.3440019  0.18985561]
 [1.91011047 1.98771158 0.39721054 3.25297437 1.21979063 0.13556017]]


**Paso 9:**
Evaluar a persona con el modelo

In [ ]:
#Para evaluar un solo persona y verificar si tuvo un accidente en la autopista o no:
XS =[[5,1,1,3,4,0]] #tiene que poner las dos variables de ingreso/entrada
mnb.predict(XS) #si nos saca un 1, se podría decir que sí tuvo accidente en autopista, si nos saca un 0, no tuvo accidente en autopista

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


array([0])

In [ ]:
ydp=mnb.predict(XD) #Hace el pronostico de los positivos y negativos
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(YD, ydp)
print(cm)

#Hay 1365 Verdaderos negativos, 215 Falsos positivos, 224 falsos negativcos y 277 verdaderos positivos
VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

Ex= (VN+VP)/(VN+FP+FN+VP)
print("La exactitud del modelo es:", Ex) #ESTO ME INDICA EL DESEMPEÑO GENERAL DEL MODELO.
#UNA CONCLUSION ES QUE PRONOSTICA MUY BIEN cerca del 80% DE LOS DATOS

Te= (FN+FP)/(VN+FP+FN+VP)
print("La tasa de error es:", Te)

Sen=VP/(VP+FN)
print("La sensibilidad del modelo es:", Sen) #Evaluamos el comportamiento del modelo calculando accidentes en la autopista

Esp=VN/(VN+FP)
print("La especificidad del modelo es:", Esp) #Evaluamos el comportamiento del modelo al predecir que no hay accidentes en la autopista

pre=VP/(VP+FP)
print("La precisión del modelo es:", pre) #Evaluamos el comportamiento del modelo identificando accidentes. Es muy similar a la sensibilidad. Pero este es más usado ya que es más preciso

pren=VN/(VN+FN)
print("La precisión negativa del modelo es:", pren) #Evaluamos el comportamiento del modelo identificando que no hay accidentes en la autopista. Es muy similar a la especificidad. Pero este es más usado ya que es mas preciso

print("Total de excatitud y tasa de error:", Ex+Te)

[[1365  215]
 [ 224  277]]
La exactitud del modelo es: 0.7890437289764536
La tasa de error es: 0.21095627102354636
La sensibilidad del modelo es: 0.5528942115768463
La especificidad del modelo es: 0.8639240506329114
La precisión del modelo es: 0.5630081300813008
La precisión negativa del modelo es: 0.8590308370044053
Total de excatitud y tasa de error: 1.0


**CONCLUSIONES**

 Hay un total de 5963 datos y 5110 (85.7%) corresponden a personas que tienen plan de salud, y 853 (14.3%)  corresponde a personas que no tienen plan de salud

De acuerdo con la información presentada, observamos que los accidentes, tanto en autopistas como fuera de ellas, tienden a ocurrir en promedio entre miércoles y jueves, sin una diferencia significativa entre ambos contextos.

Sin embargo, el nivel de impacto de los accidentes (ViolCat) es considerablemente mayor fuera de la autopista, con un promedio de 7.46, en comparación con 4.15 en las autopistas. Esta tendencia se refleja también en el tipo de accidentes (CrashType), donde los accidentes fuera de la autopista tienden a ser más severos o complejos, con un valor promedio de 3.22 frente a 2.28 en las autopistas.

Respecto a la desviación estándar y los límites, la variable Weekday muestra poca variación entre accidentes en autopistas y fuera de ellas, lo que indica que los días en que ocurren estos accidentes son bastante similares.

Al analizar el nivel de impacto (ViolCat), se observa que los accidentes fuera de la autopista no solo tienen un impacto promedio mayor, sino que también presentan una mayor variabilidad (desviación estándar más alta). Esto sugiere una mayor probabilidad de accidentes extremadamente severos fuera de la autopista, especialmente si consideramos los límites superiores. Por el contrario, en las autopistas, los accidentes tienden a ser de menor impacto y con menos variabilidad.

El tipo de accidente (CrashType) muestra mayor variabilidad fuera de la autopista, lo que refuerza la idea de que en estas áreas los accidentes pueden ser más diversos y complejos.



Hay 1365 Verdaderos negativos, 215 Falsos positivos, 224 falsos negativcos y 277 verdaderos positivos
LO que nos quiere decir y confirma que el modelo es bueno y predice correctamente la mayoría de los casos donde el evento no ocurre.

**Si evaluamos el modelo:**

La exactitud es de cerca de un 79%, lo que quiere decir que las predicciones de accidentes en la autopista o fuera de ella son muy buenas.
Como consecuencia, la tasa de error fue de un 21%

La sensibilidad es del 55%, lo que quiere decir que detecta en ese porcentaje los accidentes que en efecto ocurren en las autopistas.

Respecto a la presición positiva, esta es del 56%, lo que nos dice que tan confiable son las predicciones de datos positivos. por lo que podemos ver, no es muy preciso.

Para la especificidad, este obtuvo un valor del 86%, lo que quiere decir que detecta cerca del 90% de casos de accidentes que no ocurren en autopistas.

La precision negativa es del 85%, lo que nos da una buena idea de que el modelo es capaz de predecir muy bien los accidentes que no ocurren en las aupistas.
